src/zjs_promise.c

```diff
@@ -138,29 +117,23 @@ void zjs_fulfill_promise(jerry_value_t obj, jerry_value_t argv[], uint32_t argc)
     }
 
     jerry_get_object_native_handle(promise_obj, (uintptr_t*)&handle);
-
+    jerry_release_value(promise_obj);
     if (handle) {
-        // Put *something* here in case it never gets registered
-        if (!handle->then_set) {
-            jerry_release_value(handle->then);
-            handle->then = jerry_create_external_function(null_function);
-        }
-
-        handle->then_id = zjs_add_callback_once(handle->then,
-                                                handle->this,
+        handle->then_id = zjs_add_callback_once(ZJS_UNDEFINED,
+                                                obj,
                                                 handle,
                                                 post_promise);
 
         zjs_signal_callback(handle->then_id, argv,
                             argc * sizeof(jerry_value_t));
```

Comment:
Since you create the callback and signal it right away, the only way it will get "edited" is if the then() happens call happens in between now and when the signal is processed. Which is just the weird synchronous case, not the "normal" case, I think.